tao_compiler/mlir/xla/ral/context/stream_executor_based_impl.cc

```diff
@@ -1485,6 +1570,189 @@ void ral_conv(ExecutionContext* ctx, void* stream_handle,
   return;
 }
 
+template <typename T, int N>
+void ral_qconv(ExecutionContext* ctx, void* stream_handle,
+               MemRefType<T, N> input, MemRefType<T, N> kernel,
+               MemRefType<int32_t, 1> padding, MemRefType<float, 0> inputScales,
+               MemRefType<int32_t, 0> inputZeroPoints,
+               MemRefType<float, 1> weightScales,
+               MemRefType<int32_t, 1> weightZeroPoints,
+               MemRefType<float, 0> resultScales,
+               MemRefType<int32_t, 0> resultZeroPoints, MemRefType<T, N> output,
+               MemRefType<int32_t, 1> metadata) {
+  static_assert(N > 2, "dimension should be larger than 2");
+  if (isEmptyMemref(input) || isEmptyMemref(kernel) || isEmptyMemref(output)) {
+    TAO_VLOG(1) << "ral_conv: early return for empty tensor";
+    return;
+  }
+
+  if (TAO_VLOG_IS_ON(2)) {
+    print_memref(input, "input");
+    print_memref(kernel, "kernel");
+    print_memref(padding, "padding");
+    for (int i = 0; i < N - 2; ++i) {
+      TAO_VLOG(0) << "\tpadding for dim #" << i << ": (" << padding.data[2 * i]
+                  << ", " << padding.data[2 * i + 1] << ")";
+    }
+    print_memref(output, "output");
+    print_memref(metadata, "metadata");
+    for (int i = 0; i < 5 * N - 4; ++i) {
+      TAO_VLOG(0) << "\t#" << i << ": " << metadata.data[i];
+    }
+  }
+
+  auto gpu_driver = ctx->getDriver<GPUDriver>(GPUDriver::name());
+  float input_scale, weight_scale, result_scale;
+  gpu_driver->d2h(ctx, stream_handle, &inputScales.data[0], &input_scale,
+                  sizeof(float));
+  gpu_driver->d2h(ctx, stream_handle, &weightScales.data[0], &weight_scale,
+                  sizeof(float));
+  gpu_driver->d2h(ctx, stream_handle, &resultScales.data[0], &result_scale,
+                  sizeof(float));
+
+  float kernel_scale = input_scale * weight_scale / result_scale;
+
+#if defined(PLATFORM_ALIBABA) and defined(ENABLE_BLADE_GEMM) and \
+    !(TENSORFLOW_USE_ROCM)
+  // nchw=0123 iohw=0123
```

Comment:
Better to reuse this piece of code with normal ral_conv.

Reply:
Done